File: dev/utils/DayTradeBook.py
# -*- coding: utf-8 -*-
"""
Created on Fri Nov 12 19:08:16 2021

@author: infomax
"""


import pandas as pd
import numpy as np
import datetime
import os
from tqdm import tqdm
import sys
sys.path.append('D:\\dev\\kbsecurities\\dev\\utils')
import util

class DayTradeBook:
    
    """
    한 전략에 대해 해당일의 매매에 대한 기록을 담는 class
        - 상품에 대한 기본 정보도 같이 담는다
        - PLOTTING 기능도 탑재?
    
    2021.11.12
    """
    
    def __init__(self, date, db_table):
        
        if db_table[:3] == 'lkt':
            self.product = 'lktbf'
            self.tick_value = 0.01
            self.krw_commission_per_contract = 250 # 원래 248.95원/계약
            self.krw_value_1pt = 10**6
        elif db_table[:3] == 'ktb':
            self.product = 'ktbf'
            self.tick_value = 0.01
            self.krw_commission_per_contract = 250 # 원래 225.5원/계약
            self.krw_value_1pt = 10**6
        elif db_table[:3] == 'usd':
            self.product = 'usdkrw'
            self.tick_value = 0.1
            self.krw_commission_per_contract = 60 # 원래 57.9원/계약
            self.krw_value_1pt = 10**4
        else:
            raise NameError('Unexpected product')
            
        self.tick_conversion = 1 / self.tick_value
                
        self.date = date
        self.yday = util.date_offset(date, -1)
        
        self.yday_close = util.getDdayOHLC(self.yday, table=self.product + '_day')['close']
        self.yday_lo = util.getDdayOHLC(self.yday, table=self.product + '_day')['low']
        self.yday_hi = util.getDdayOHLC(self.yday, table=self.product + '_day')['high']
        
        self.siga_time = util.getMktTime(date)['start']
        
        self.book = pd.DataFrame(
            columns=['signal_time',
                     'signal_price',
                     'trade_time',
                     'trade_price',
                     'trade_qty',
                     'trade_type',
                     'remark',
                     'pos_id',
                     ])
        
        
        
    def logTrade(self, trade_time, trade_price, trade_qty, trade_type, 
                 signal_time=None, signal_price=None, remark=None):
        if signal_time == None: signal_time = trade_time
        if signal_price == None: signal_price = trade_price
        
        if trade_type == 'ini':
            if self.book.empty:
                pos_id = 1
            else:
                pos_id = self.book['pos_id'].iloc[-1] + 1
        else:
            pos_id = self.book['pos_id'].iloc[-1]
            
        row =[signal_time, signal_price, 
              trade_time, trade_price, trade_qty, 
              trade_type, remark, pos_id]
        self.book.loc[len(self.book)] = row
        print(trade_time, trade_price, trade_qty, trade_type, remark)
        pass
    
    def exitOpenPosition(self, trade_time, mid_price, remark, signal_price=None):
        open_position_qty = self.getOpenPositionQty()
        
        if open_position_qty == 0:
            raise ValueError("No Open Position")
        else:
            if open_position_qty > 0:
                self.logTrade(trade_time = trade_time, 
                              trade_price = mid_price - 0.5*self.tick_value, 
                              trade_qty = -open_position_qty,
                              trade_type='ext',
                              remark = remark,
                              signal_price=signal_price)
            else:
                self.logTrade(trade_time = trade_time, 
                              trade_price = mid_price + 0.5*self.tick_value, 
                              trade_qty = -open_position_qty, 
                              trade_type='ext',
                              remark = remark,
                              signal_price=signal_price)
        pass
    
    def addDatetimeColumn(self):
        if not self.book.empty:
            self.book['datetime'] = self.book['trade_time'].apply(lambda x: x + pd.Timestamp(self.date))
        pass
        
    def getOpenPositionQty(self):
        return self.book['trade_qty'].sum()
    
    def _getOpenPositionStartIndex(self):
        if self.book.empty:
            raise ValueError("No Trade yet")
        else:
            if not 'ext' in self.book.trade_type.values: # ext trade 한번도 없었음
                return 0
            elif self.book.trade_type.values[-1] == 'ext':
                raise ValueError("No Open Position")
            else:
                last_exit_index = self.book[self.book['trade_type'] == 'ext'].index[-1]
            return last_exit_index+1
    
    def _getOpenPositionTradeQtyArray(self):
        i_open = self._getOpenPositionStartIndex()
        return np.array(self.book[i_open:]['trade_qty'])
    
    def _getOpenPositionTradePriceArray(self):
        i_open = self._getOpenPositionStartIndex()
        return np.array(self.book[i_open:]['trade_price'])
    
    def getLastTradePrice(self):
        if self.book.empty:
            raise ValueError("No trade yet")
        else:
            return self.book['trade_price'].iloc[-1]
        
    def getLastTradeQty(self):
        if self.book.empty:
            raise ValueError("No trade yet")
        else:
            return self.book['trade_qty'].iloc[-1]
    
    def isMovedEnoughFromLastTrade(self, price, min_price_move_required):
        # mean reversion 관점에서 최근거래대비 가격 개선 여부를 체크 
        if self.book.empty: # 거래가 없는 경우
            return True
        elif self.getOpenPositionQty() == 0: # 거래는 있었는데 Open은 없는 경우
            return True
        else: # open position이 있는 경우
            # 마지막이 숏인 경우 더 강해져야 add
            if  self.getLastTradeQty() < 0:
                if price - self.getLastTradePrice() > min_price_move_required:
                    return True
            # 마지막이 롱인 경우 더 밀려야 add
            elif self.getLastTradeQty() > 0:
                if price - self.getLastTradePrice() < -min_price_move_required:
                    return True
            else: 
                return False
    
    def getLastClosedPositionInfo(self):
        if not 'ext' in self.book.trade_type.values: # ext trade 한번도 없었음
            raise ValueError("No Closed position yet")
        else:
            last_exit_index = self.book[self.book['trade_type'] == 'ext'].index[-1]
            last_pos_id = self.book.loc[last_exit_index]['pos_id']
            last_pos_book = self.book[self.book['pos_id'] == last_pos_id]
            
            last_pos_direction = 'short' if last_pos_book.iloc[0]['trade_qty'] < 0 else 'long'
            last_pos_remark = last_pos_book.iloc[-1]['remark']
            
            return last_pos_direction, last_pos_remark
            
        
    
    def getOpenPnl(self, price, unit='krw'):
        """
        returns open position pnl in unit ('krw' or 'tick')
        """
        if self.getOpenPositionQty() == 0:
            raise ValueError("No Open Position")
        else:
            qty_arr = self._getOpenPositionTradeQtyArray()
            price_arr = self._getOpenPositionTradePriceArray()
            sumproduct = sum(qty_arr * (price - price_arr))
            commission = sum(abs(qty_arr)) * self.krw_commission_per_contract
            
            if unit == 'krw':
                return round(self.krw_value_1pt * sumproduct - commission, 0)
            
            elif unit == 'tick': # without commission
                return round(self.tick_conversion * sumproduct / sum(qty_arr), 2)
            
    def getCumPl(self, price=None, unit='krw'):
        """
        returns cumulative pnl in unit ('krw' or 'tick')
        """
        if self.book.empty:
            return 0
        else:
            if price == None:
                price = self.book['trade_price'].iloc[-1]
            qty_arr = np.array(self.book.trade_qty)
            price_arr = np.array(self.book.trade_price)
            sumproduct = sum(qty_arr * (price - price_arr))
            commission = sum(abs(qty_arr)) * self.krw_commission_per_contract
            
            if unit == 'krw':
                return round(self.krw_value_1pt * sumproduct - commission, 0)
            
            elif unit == 'tick': # without commission
                return round(self.tick_conversion * sumproduct / sum(qty_arr), 2)

    def __repr__(self):
        return self.book.__repr__()


# date = datetime.date(2021, 1, 14)
# b = DayTradeBook(date, 'lktbf_10sec')

# tl = []
# for i in range(8):
#     tl.append(datetime.time(9, i+1))

# prl = []
# for i in range(8):
#     prl.append(101.01 + i/100)

# ql = [20, -20, 20, 20, -40, -20, 20, 20]

# tyl = ['ini', 'ext', 'ini', 'add', 'ext', 'ini', 'ext', 'ini']

# for i in range(8):
#     b.logTrade(tl[i], prl[i], ql[i], tyl[i])

# print(b)
# print(b.getCumPl(101.10))


# b2 = DayTradeBook(date, 'lktbf_10sec')
# b2.logTrade(tl[0], prl[0], ql[0], 'ini')
# b2.logTrade(tl[0], prl[0], ql[0], 'add')

# print(b2.getOpenPnl(101.10))

# df = b.book
# df2 = b2.book
# df0 = DayTradeBook(date, 'lktbf_10sec').book
# dfext = df.drop(index=7)

        
    
    
    
    
    
        
        
        
            

File: prd/APISample/DayTradeBook.cpp
#include "stdafx.h"
#include "DayTraderBook.h"
using namespace std;

DayTraderBook::DayTraderBook() {

	CString jongmok = "LKTBF";
	if (jongmok == "KTBF" || jongmok == "LKTBF") {
		product = jongmok;
		tick_value=0.01;
		krw_commission_per_contract = 250;
		krw_value_1pt = 1000000;
	} else if (jongmok == "USD") {
		product = jongmok;
		tick_value = 0.1;
		krw_commission_per_contract = 60;
		krw_value_1pt = 10000;
	} else {
		throw ERROR("Unexpected product");
	}
	tick_conversion = 1 / tick_value;
	book.trade_num = 0;
}

DayTraderBook::DayTraderBook(CString jongmok) {
	if (jongmok == "KTBF" || jongmok == "LKTBF") {
		product = jongmok;
		tick_value=0.01;
		krw_commission_per_contract = 250;
		krw_value_1pt = 1000000;
	} else if (jongmok == "USD") {
		product = jongmok;
		tick_value = 0.1;
		krw_commission_per_contract = 60;
		krw_value_1pt = 10000;
	} else {
		throw ERROR("Unexpected product");
	}
	tick_conversion = 1 / tick_value;
	book.trade_num = 0;
}

void DayTraderBook::logTrade(CString trade_time, float trade_price, int trade_qty, CString trade_type, CString signal_time, float signal_price, CString remark) {

	if (signal_time == "") {signal_time = trade_time;}
	if (signal_price == 0) {signal_price = trade_price;}

	int pos_id=1;
	if (trade_type == "ini") {
		if (book.trade_num == 0) {
			pos_id = 1;
		} else {
			pos_id = book.pos_id[book.trade_num] +1;
		}
	} else {
		pos_id = book.pos_id[book.trade_num];
	}

	book.signal_time[book.trade_num] = signal_time;
	book.signal_price[book.trade_num] = signal_price;
	book.trade_qty[book.trade_num] = trade_qty;
	book.trade_type[book.trade_num] = trade_type;
	book.trade_time[book.trade_num] = trade_time;
	book.trade_price[book.trade_num] = trade_price;
	book.remark[book.trade_num] = remark;
	book.pos_id[book.trade_num] = pos_id;
	
	book.trade_num++;
}

void DayTraderBook::exitOpenPosition(CString trade_time, float mid_price, CString remark, float signal_price) {
	int open_position_qty = getOpenPositionQty();
	if (open_position_qty == 0) {
		throw ERROR("No open position");
	} else {
		if (open_position_qty >0 ) {
			logTrade(trade_time, mid_price-0.5*tick_value, -open_position_qty, "ext", "", 0,remark);
		} else {
			logTrade(trade_time, mid_price+0.5*tick_value, -open_position_qty, "ext", "", 0, remark);
		}
	}
}

int DayTraderBook::getOpenPositionQty() {
	int sumQty = 0;
	for (int i = 0; i < book.trade_num; i++) {
		sumQty += book.trade_qty[i];
	}
	return sumQty;
}

int DayTraderBook::_getOpenPositionStartIndex() {
	if (book.trade_num == 0) {
		throw ERROR("No trade yet"); // no open position
	} else { 
		int idx = 0;
		for (int i = book.trade_num-1; i >0 ;i--) {
			if (book.trade_type[i] =="ext") {
				idx = i;
				break;
			}
		}

		if (idx == 0) { // non exist
			return 0;
		} else if (idx == book.trade_num-1) {
			throw ERROR("No open position");  //no open position
		} else {
			return idx+1; // openposition 이 있고 정상적인 경우
		}

		return -1;
	}
}

//bool DayTraderBook::isMovedEnoughFromLastTrade(float price, float min_price_move_required) {
//	if (book.trade_num == 0) {
//		return true;
//	} else {
//		float abs = (price - book.trade_price[book.trade_num-1] > 0) ? price - book.trade_price[book.trade_num-1] : - price + book.trade_price[book.trade_num-1];
//		if (abs > min_price_move_required) {
//			return true;
//		} else {
//			return false;
//		}
//	}
//}
int DayTraderBook::getCurQty() {
	if(book.trade_num==0) {
		return 0;
	}
	int qty = 0;
	for(int i =0; i< book.trade_num; i++) {
		qty += book.trade_qty[i];
	}
	return qty;
}

float DayTraderBook::getOpenPnl(float price, CString unit) {
	if (getOpenPositionQty() == 0 ) { 
		throw ERROR("No open position"); // no open position
	} else {
		float sum_product = 0;
		float commission = 0;
		for (int i = _getOpenPositionStartIndex(); i< book.trade_num; i++ ) {
			sum_product += book.trade_qty[i] * book.trade_price[i];
			commission += krw_commission_per_contract * ((book.trade_qty[i] >0)? book.trade_qty[i] : -book.trade_qty[i]);
		}

		if (unit =="krw") {
			return int(krw_value_1pt * sum_product-commission);
		} else if (unit =="tick") { // without commission
			int tmp = tick_conversion * sum_product / (float)getOpenPositionQty() * 100;
			return (float)tmp / 100;
		}
	}
}

float DayTraderBook::getCumPl(float price, CString unit) {
	if (book.trade_num == 0) {
		return 0;
	} else {
		float sum_product = 0;
		float commission = 0;
		for (int i = 0; i< book.trade_num; i++ ) {
			sum_product += book.trade_qty[i] * (price - book.trade_price[i]);
			commission += krw_commission_per_contract * ((book.trade_qty[i] >0)? book.trade_qty[i] : -book.trade_qty[i]);
		}
		if (unit == "krw") {
			return int(krw_value_1pt * sum_product - commission);
		} else if (unit =="tick") {
			int tmp = tick_conversion * sum_product / (float)getOpenPositionQty() * 100;
			return (float)tmp / 100;
		}
	}
}